core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java

```diff
@@ -109,18 +112,34 @@ protected abstract class BaseEqualityDeltaWriter implements Closeable {
     private final StructProjection structProjection;
     private RollingFileWriter dataWriter;
     private RollingEqDeleteWriter eqDeleteWriter;
-    private SortedPosDeleteWriter<T> posDeleteWriter;
+    private FileWriter<PositionDelete<T>, DeleteWriteResult> posDeleteWriter;
     private Map<StructLike, PathOffset> insertedRowMap;
 
     protected BaseEqualityDeltaWriter(StructLike partition, Schema schema, Schema deleteSchema) {
+      this(partition, schema, deleteSchema, DeleteGranularity.PARTITION);
+    }
+
+    protected BaseEqualityDeltaWriter(
+        StructLike partition,
+        Schema schema,
+        Schema deleteSchema,
+        DeleteGranularity deleteGranularity) {
```

Comment:
I am not convinced we should make it configurable when writing equality deltas. Our original plan was to always generate file-scoped position deletes when writing Flink CDC. Take a look at [this](https://github.com/apache/iceberg/pull/9384#discussion_r1439722314) discussion.

Reply:
What is the current experience is Spark with this opinion? Are we planning to move the Spark deletes to file granularity?

I would like to hear @stevenzwu's opinion in this regard, but I would be happy to eliminate a configuration possibility 😄

Reply:
We don't have a lot of input from the Spark side as it was shipped in 1.5 for the first time. There, the situation is also very different as the number of deletes is different.

@chenjunjiedada, you faced the concurrent compaction issue in the past. Would you have a chance to try this PR and see how it performs for your use cases?

@stevenzwu @rdblue @RussellSpitzer, would love to hear your thoughts as well.

Reply:
We discussed a bit on slack. I think no-conf is good, in my mind if there was a benefit from having the two modes as an option it's something flink should be choosing at runtime rather than a user defined config. I think this is also mostly a perf difference so I wouldn't mind changing the behavior from the previous behavior in those edge cases with enough data to write multiple files in the same partition in one batch.

Reply:
I am also in favor of no config. It is a "good" internal impl/perf change. In the data file rollover case, one smaller delete file every data file is better than one larger delete file for multiple data files.

Reply:
Seems like everyone is in favor of not exposing the config then.

Reply:
I love this 😄 
The PR is sooo much smaller!

@stevenzwu: one final round of review?